### ISSUE TITLE
Tailscale breaks DNS-over-TLS on Android

### ISSUE THREAD

**Issue by michaelclarkcuadrado** [tailscale/tailscale#915]
**Describe the bug**
If a custom DNS-over-TLS ("Private DNS") server is configured, enabling tailscale will stop all DNS requests and show a "couldn't connect" in the android settings app.

**To Reproduce**
1. Disable the tailscale VPN. 
2. In Android setting's "WiFi and network" section, set the private dns to a specific server, not "automatic". I tested with values of both "dns.quad9.net" and "dns.nextdns.io". 
3. Observe that dns works as expected.
4. Activate tailscale.
5. Observe that dns requests fail, and that private dns' settings are labelled "couldn't connect"

Note that the "automatic" setting, for Google's DNS-over-TLS server, works correctly.

**Expected behavior**
DNS-over-TLS servers should be accessible even when running tailscale.

**Screenshots**
![Screenshot_20201111-121933](https://user-images.githubusercontent.com/15159826/98860683-e84b5c00-2418-11eb-9e0c-01ff545a6bf0.jpg)

**Version information:**
 - Device: OnePlus Phone
 - OS: Android
 - OS version: 10.
 - Tailscale version: version 1.2.2-t76c2982d8-g39dfd8495

**Additional context**

**Comment by bradfitz** [comment]
Can you also share your Tailscale IP address?

**Comment by michaelclarkcuadrado** [comment]
The tailscale IP of the device is 100.73.211.12

**Comment by normanr** [comment]
I see the same thing when trying to use dns.google on 100.66.241.34.

If I use socat to ssl listen on port 853 (and tcp connect to port 53) on the dns server I configured in Tailscale, then Automatic mode seems to use it. Manual mode works too as long as I change my preferred Private DNS server name to match the certificate I'm using.

Just a plain forwarding of port 853 to the upstream dns server also works in Automatic mode, but not with a provided hostname. Could it be because Tailscale forces all DNS traffic to 100.100.100.100 which isn't listening on port 853?

**Comment by normanr** [comment]
note: I don't think that "Automatic" mode uses Google's DNS-over-TLS server, but rather it checks if the currently configured DNS server supports DNS-over-TLS and uses it if it's available.

If you want to use Google's DNS-over-TLS server then I think you have to explicitly enter dns.google as your Private DNS provider.

**Comment by DentonGentry** [comment]
I'm running into this as well: it initially manifested only when connecting via LTE, not Wi-Fi. A subset of domains would not resolve, notably *.appspot.com (such as gfblip.appspot.com).

At the time I did not have Private DNS configured in Settings, so Android would default to trying DNS-over-TLS if the DNS server supported it. When connecting via Wifi I'm using a DNS server within the home, which doesn't implement DNS-over-TLS and also recursively resolves on behalf of clients. As Android never tries to use DNS-over-TLS it doesn't run into a problem with Tailscale on when connected via Wifi.

When connecting via LTE Android uses whatever the carrier provides, which I'm guessing is not a recursive resolver so Android has to connect to the DNS servers for various domains. DNS for appspot.com supports DNS-over-TLS, Android with Tailscale on isn't connecting to it.

Configuring Private DNS shows the problem more clearly, since it tries to use DNS-over-TLS all the time. The attached screenshots with Private DNS set to "dns.google" and tailscale on and off.
![private_dns_tailscale_off](https://user-images.githubusercontent.com/1140756/101024460-f7f62600-3528-11eb-8390-032353086a5c.png)
![private_dns_tailscale_on](https://user-images.githubusercontent.com/1140756/101024467-f9bfe980-3528-11eb-9a1c-af2a8aa30351.png)

Tailscale's Magic DNS does not impact this issue, turning MagicDNS on and off does not make a difference.

With PrivateDNS configured I can produce the problem on Wifi, which makes it easier to get packet captures from the router. I'll try to do that next.

**Comment by DentonGentry** [comment]
TL;DR: have not managed to produce similar behavior with DNS-over-TLS + Ubuntu + Tailscale (which would have been nice as it is much easier to debug).

I enabled DNS-over-TLS on Ubuntu 20.04:

```
$ cat /etc/systemd/resolved.conf
[Resolve]
DNS=8.8.8.8
FallbackDNS=8.8.4.4
#DNS=
#FallbackDNS=
Domains=~
#LLMNR=no
#MulticastDNS=no
#DNSSEC=yes
#DNSOverTLS=opportunistic
DNSOverTLS=yes
#Cache=yes
#DNSStubListener=yes
#ReadEtcHosts=yes
```

```
$ sudo systemctl daemon-reload
$ sudo systemctl restart systemd-networkd
$ sudo systemctl restart systemd-resolved
```

Packet captures on the router show no traffic from this Ubuntu host to port 53 any more, only port 853. If I change back to DNSOverTLS=no, the router shows traffic from this Ubuntu host to port 53 again and no traffic to 853.

`tailscale status` shows the Ubuntu system is connected. I think this is the circumstance where Android fails to resolve addresses, however no such issue is noted on the Ubuntu system. One of the hosts which notably failed on Android is gfblip.appspot.com, but on Ubuntu there is no problem resolving gfblip.appspot.com.

I'll poke around a bit more with Ubuntu, but if it doesn't seem likely to produce a problem I'll look for a reasonable way to debug on Android.

**Comment by DentonGentry** [comment]
Noticed something important while out getting take-out.

TL;DR: MagicDNS triggers the problem to occur. Once the device is in the problem state, disabing and re-enabling MagicDNS makes no difference.

With a freshly booted Android device, I can enable Tailscale and:
+ switch to LTE yet still resolve gfblip.appspot.com
+ turn on PrivateDNS to use DNS-over-TLS for all lookups

Everything works fine. These are the two scenarios which were failing before.

When I enable MagicDNS for the first time:
+ on LTE, *some* domains like *.appspot.com become non-resolvable
+ with PrivateDNS on, all name resolution fails

Once in the failing state, disabling MagicDNS makes no difference. Name resolution on the device continues to fail.

In https://github.com/tailscale/tailscale/issues/915#issuecomment-738003635 I said that MagicDNS does not impact the issue. My device was already in the failed state at that point, and I definitely had turned MagicDNS on earlier that day.

Initial speculation was that maybe the client could no longer resolve "tailscale.com" and therefore lost the ability to contact Control, but I see it checking in at https://login.tailscale.com/admin/machines. The problem appears to be more subtle than that.

**Comment by bradfitz** [comment]
Phew! That's less mysterious now. 😅

**Comment by DentonGentry** [comment]
Trying again to reproduce on Ubuntu, with the added detail of enabling MagicDNS to trigger the problem, I still don't see the failure on Ubuntu.

With tailscale up and MagicDNS enabled, /etc/resolv.conf contains:
```
$ cat /etc/resolv.conf
# resolv.conf(5) file generated by tailscale
# DO NOT EDIT THIS FILE BY HAND -- CHANGES WILL BE OVERWRITTEN

nameserver 100.100.100.100
search example.com.beta.tailscale.net.
``` 

With tailscale still up and after disabling MagicDNS:
```
$ cat /etc/resolv.conf
# resolv.conf(5) file generated by tailscale
# DO NOT EDIT THIS FILE BY HAND -- CHANGES WILL BE OVERWRITTEN

nameserver 8.8.8.8
nameserver 8.8.4.4
```

These are both what we would expect, and DNS lookups are working without issue in both cases.

With MagicDNS disabled I can run wireshark on the Ethernet interface and see DNS-over-TLS traffic on port 853, also as expected, and these queries get responses. With MagicDNS enabled I can't see the DNS queries: even if they traversed the tailscale0 interface they would be Wireguard-encrypted, but as I understand it they get resolved locally and don't get sent upstream anyway.

So unfortunately it looks like this has to be debugged on the Android device, its resolver is quite different and the problem appears to be specific to something about the platform.

**Comment by DentonGentry** [comment]
Using adb logcat after enabling PrivateDNS with MagicDNS enabled:
`12-14 21:28:04.116  1466  1819 D DnsManager: sendDnsConfigurationForNetwork(106, [100.100.100.100], [example.com.beta.tailscale.net.], 1800, 25, 8, 64, 0, 0, dns.google, [])`

`12-14 21:28:04.355  1466  1819 D DnsManager: sendDnsConfigurationForNetwork(104, [fd00:976a::9, fd00:976a::10], [], 1800, 25, 8, 64, 0, 0, dns.google, [2001:4860:4860::8888, 2001:4860:4860::8844, 8.8.8.8, 8.8.4.4])`

The first log message for network #106 is clearly for tailscale0 as it sets a search domain which ends in tailscale.net. The second log message is for a different interface, probably Wi-Fi.


The log message comes from https://android.googlesource.com/platform/frameworks/base/+/master/services/core/java/com/android/server/connectivity/DnsManager.java. I don’t see a way to link to a specific version of the file on android.googlesource.com. As of 15 Dec 2020 the file hash is c70bb080b0b1545d29f34d2a9df73adb0db27a48 and the code for that log message is on line 389:

```
        Log.d(TAG, String.format("sendDnsConfigurationForNetwork(%d, %s, %s, %d, %d, %d, %d, "
                + "%d, %d, %s, %s)", paramsParcel.netId, Arrays.toString(paramsParcel.servers),
                Arrays.toString(paramsParcel.domains), paramsParcel.sampleValiditySeconds,
                paramsParcel.successThreshold, paramsParcel.minSamples,
                paramsParcel.maxSamples, paramsParcel.baseTimeoutMsec,
                paramsParcel.retryCount, paramsParcel.tlsName,
                Arrays.toString(paramsParcel.tlsServers)));
```

The last parameter is the set of DNS-over-TLS servers to use, and for tailscale0 it is empty. I suspect this is the reason for most symptoms: Android has been set to only use DNS-over-TLS, and there are no DNS-over-TLS servers available.

----

Android Tailscale app uses [VpnService.Builder addDnsServer](https://developer.android.com/reference/android/net/VpnService.Builder#addDnsServer(java.lang.String)) to add the DNS servers, and I see no mention of DNS-over-TLS in the VpnService.Builder documentation.

The code for VpnService.Builder.addDnsServer is at https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/net/VpnService.java (as of 15 Dec 2020 the file hash is c70bb080b0b1545d29f34d2a9df73adb0db27a48 and the relevant code is on line 632:

```
@NonNull
public Builder addDnsServer(@NonNull InetAddress address) {
    if (address.isLoopbackAddress() \|\| address.isAnyLocalAddress()) {
        throw new IllegalArgumentException("Bad address");
    }
    if (mConfig.dnsServers == null) {
        mConfig.dnsServers = new ArrayList<String>();
    }
    mConfig.dnsServers.add(address.getHostAddress());
        return this;
    }
 }
```

**Comment by DentonGentry** [comment]
Noting:
+ [issue report that Google Fi VPN breaks Android Private DNS](https://www.reddit.com/r/ProjectFi/comments/ag273s/fi_vpn_breaks_googles_dnsovertls/), with the most visible system that dns.google says, "Couldn't connect." This is the same symptom seen here.
+ [Cisco VPN client for Android bug report](https://quickview.cloudapps.cisco.com/quickview/bug/CSCvr47450) that DNS resolution fails for the VPN tunnel for 1) Android 10, 2) Private DNS is Enabled or Private DNS is in Automatic mode and the public network DNS servers are capable of DNS over TLS, and 3) Split include tunnel configuration. This also sounds very much like the symptoms seen here. Unfortunately seeing the details of the Cisco bug report or any followups require a paid Cisco support contract, which I don't have.

Looking through the [VpnService code](https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/net/VpnService.java) there is no mention of DNS-over-TLS. I feel a sense of foreboding, that Private DNS and Android VPNService simply don't work together.

Summarizing the current state of the investigation: so far as I can see, Android makes no attempt to send packets to port 853. Neither logging added in wgengine/userspace.go nor wireshark running on the Wi-Fi router show any packets to port 853. I think whatever is going wrong happens before the point that Android even tries to validate the DNS-over-TLS servers, and it decides that it has no suitable DNS servers. If strict mode Private DNS is enabled, this means that Android cannot resolve any domain names at all.

**Comment by DanielJoyce** [comment]
@DentonGentry Seeing this behavior too, confirmed that disabling magic dns and removing the nameservers fixes it.

**Comment by DanielJoyce** [comment]
Just wanted to clarify, is there a reason we are configuring the private dns feature of android instead of adding a dns entry to the VPNService builder? Am I missing something? Not familiar with the android code. Does VpnService not support Dns over tls? Interplay issue? Something ekse?

**Comment by DentonGentry** [comment]
So far as I can tell the problem is an incompatibility within Android between the VpnService addDnsServer\(\) and the PrivateDNS feature. If DNS servers have been added using VpnServiceBuilder.addDnsServer\(\), then PrivateDNS does not work and fails before it even attempts to send DNS-over-TLS packets to port 853. The basis for this belief is that [Google Fi VPN and Cisco VPN experience the same symptoms](https://github.com/tailscale/tailscale/issues/915#issuecomment-745660140).

My guess would be that ExpressVPN does not add call addDnsServer\(\) and does not add DNS servers of its own. It looks like it is never used as a split VPN: its primarily goal is to protect privacy and anonymity, tunneling everything back to the VPN server would be sensible. So it could leave Android with whatever DNS servers it already had, allow the DNS packets to be tunneled back to the VPN headend, and handle the DNS packets from there.

**Comment by DentonGentry** [comment]
Noting from an earlier discussion: turning MagicDNS into a recursive, caching nameserver might be a solution to allow its use. That way, even in opportunistic mode the Android client won't try to use DNS-over-TLS. We could, in turn, make MagicDNS implement opportunistic DNS-over-TLS when recursively resolving and (probably) not trigger the problem in Android.

**Comment by DanielJoyce** [comment]
I've used other products kinda like tailscale, and don't have the same problem, but I also think they aren't using private dns. So that makes sense.

**Comment by DentonGentry** [comment]
I've gotten AOSP to build and run within the Android Studio emulator, allowing more logging to be added. When Private DNS is enabled in its always-on strict mode, where one supplies a hostname like dns.google, this is what happens:

```
12-28 17:59:00.458   777  1943 D NetworkMonitor/102: Strict mode hostname resolution 329ms OK 8.8.4.4,8.8.8.8
12-28 17:59:00.458   777  1943 D NetworkMonitor/102: Strict mode hostname resolution OK: dns.google [/2001:4860:4860::8844, /2001:4860:4860::8888, /8.8.4.4, /8.8.8.8]
12-28 17:59:00.459   478   592 D ConnectivityService: EVENT_PRIVATE_DNS_CONFIG_RESOLVED: [102 WIFI|VPN]
12-28 17:59:00.459   478   592 D ConnectivityService: ConnectivityService updatePrivateDns: [102 WIFI|VPN]
12-28 17:59:00.459   478   592 W DnsManager: updatePrivateDns(102, PrivateDnsConfig{true:dns.google/[/2001:4860:4860::8844, /2001:4860:4860::8888, /8.8.4.4, /8.8.8.8]})

12-28 17:59:00.459   478   592 D ConnectivityService: Setting DNS servers for network 102 to [/100.100.100.100]
12-28 17:59:00.459   478   592 D ConnectivityService: updateDnses calling noteDnsServersForNetwork for 102
12-28 17:59:00.459   478   592 W DnsManager: sendDnsConfigurationForNetwork 102 strictMode=true useTls=true
12-28 17:59:00.459   478   592 D DnsManager: sendDnsConfigurationForNetwork(102, [100.100.100.100], [example.com.beta.tailscale.net.], 1800, 25, 8, 64, 0, 0, dns.google, [])
```

The first few lines are from just before the Tailscale app calls VpnServiceBuilder.addDnsServer(), and shows Android adding the 8.8.8.8 et al servers with strict mode enabled. Android had successfully made a connection to those servers to validate DNS-over-TLS support.

The second set of log lines is from the call to VpnServiceBuilder.addDnsServer(), which has no field for DNS-over-TLS and no way for the app to express "leave the DNS-over-TLS servers alone." The last parameter in the last log line is the list of DNS-over-TLS servers, and it is \[\]. Crucially though, strict mode is still enabled because the previous verification succeeded.

So now Android has been set to require TLS, but has no DNS-over-TLS servers. Attempts to use non-TLS servers like MagicDNS are blocked by strict mode:

```
12-28 15:36:50.361  1864  1912 I com.tailscale.ipn: 16.1M/149.2M Accept: UDP{100.100.100.100:53 > 100.75.13.98:43533} 112 udp cached
12-28 15:36:50.364   799   799 D StrictMode: StrictMode policy violation: android.os.strictmode.UntaggedSocketViolation: Untagged socket detected; use TrafficStats.setThreadSocketTag() to track all network usage
12-28 15:36:50.364   799   799 D StrictMode:    at android.os.StrictMode.onUntaggedSocket(StrictMode.java:2182)
12-28 15:36:50.364   799   799 D StrictMode:    at com.android.server.NetworkManagementSocketTagger.tag(NetworkManagementSocketTagger.java:82)
12-28 15:36:50.364   799   799 D StrictMode:    at libcore.io.BlockGuardOs.tagSocket(BlockGuardOs.java:54)
12-28 15:36:50.364   799   799 D StrictMode:    at libcore.io.BlockGuardOs.socket(BlockGuardOs.java:374)
12-28 15:36:50.364   799   799 D StrictMode:    at libcore.io.ForwardingOs.socket(ForwardingOs.java:216)
12-28 15:36:50.364   799   799 D StrictMode:    at android.system.Os.socket(Os.java:640)
12-28 15:36:50.364   799   799 D StrictMode:    at android.net.util.DnsUtils.findSrcAddress(DnsUtils.java:171)
12-28 15:36:50.364   799   799 D StrictMode:    at android.net.util.DnsUtils.rfc6724Sort(DnsUtils.java:146)
12-28 15:36:50.364   799   799 D StrictMode:    at android.net.DnsResolver$InetAddressAnswerAccumulator.maybeReportAnswer(DnsResolver.java:283)
12-28 15:36:50.364   799   799 D StrictMode:    at android.net.DnsResolver$InetAddressAnswerAccumulator.onAnswer(DnsResolver.java:305)
12-28 15:36:50.364   799   799 D StrictMode:    at android.net.DnsResolver$InetAddressAnswerAccumulator.onAnswer(DnsResolver.java:251)
12-28 15:36:50.364   799   799 D StrictMode:    at android.net.DnsResolver.lambda$registerFDListener$8(DnsResolver.java:511)
12-28 15:36:50.364   799   799 D StrictMode:    at android.net.-$$Lambda$DnsResolver$hIO7FFv0AXN6Nj0Dzka-LD8S870.run(Unknown Source:8)
12-28 15:36:50.364   799   799 D StrictMode:    at com.android.networkstack.util.DnsUtils.lambda$getAllByName$1(DnsUtils.java:135)
12-28 15:36:50.364   799   799 D StrictMode:    at com.android.networkstack.util.-$$Lambda$DnsUtils$yTVtff-6rgd-poAiwKNFKah2CA4.execute(Unknown Source:0)
12-28 15:36:50.364   799   799 D StrictMode:    at android.net.DnsResolver.lambda$registerFDListener$9(DnsResolver.java:493)
12-28 15:36:50.364   799   799 D StrictMode:    at android.net.-$$Lambda$DnsResolver$kxKi6qjPYeR_SIipxW4tYpxyM50.onFileDescriptorEvents(Unknown Source:12)
12-28 15:36:50.364   799   799 D StrictMode:    at android.os.MessageQueue.dispatchEvents(MessageQueue.java:293)
12-28 15:36:50.364   799   799 D StrictMode:    at android.os.MessageQueue.nativePollOnce(Native Method)
12-28 15:36:50.364   799   799 D StrictMode:    at android.os.MessageQueue.next(MessageQueue.java:335)
12-28 15:36:50.364   799   799 D StrictMode:    at android.os.Looper.loop(Looper.java:183)
```

Therefore all DNS lookups fail until Private DNS is turned off.

**Comment by bradfitz** [comment]
Nice to see [my android.os.StrictMode](https://android-developers.googleblog.com/2010/12/new-gingerbread-api-strictmode.html) still being used! 😄

**Comment by bradfitz** [comment]
I'd propose trying to run a DNS-over-HTTPs server on 100.100.100.100 instead of the classic UDP DNS server we run, but surely (hopefully) Android is checking the TLS cert validity. They have a cert for 8.8.8.8, etc:

```
bradfitz@tsdev:~$ curl -I -v https://8.8.8.8/
* Expire in 0 ms for 6 (transfer 0x563724770bf0)
*   Trying 8.8.8.8...
* TCP_NODELAY set
* Expire in 200 ms for 4 (transfer 0x563724770bf0)
* Connected to 8.8.8.8 (8.8.8.8) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: none
  CApath: /etc/ssl/certs
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN, server accepted to use h2
* Server certificate:
*  subject: C=US; ST=California; L=Mountain View; O=Google LLC; CN=dns.google
*  start date: Nov 10 14:28:22 2020 GMT
*  expire date: Feb  2 14:28:21 2021 GMT
*  subjectAltName: host "8.8.8.8" matched cert's IP address!
*  issuer: C=US; O=Google Trust Services; CN=GTS CA 1O1
*  SSL certificate verify ok.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x563724770bf0)
> HEAD / HTTP/2
> Host: 8.8.8.8
> User-Agent: curl/7.64.0
> Accept: */*
```

Us getting one for MagicDNS at 100.100.100.100 would be tricky.

**Comment by normanr** [comment]
Note that if you configure Private DNS to use a server that is accessible via a Tailscale subnet route, then I think it works correctly.

So a way to work around the Android bug/feature might be to generate a virtual "accepted route" for the Private DNS server on the client (if it's possible to retrieve the current setting from Android), but then don't route that traffic via Wireguard if the route wasn't actually advertised anywhere on the Tailscale network. The Tailscale client will have to proxy the TCP connection "around" the Android "VPN split" in order to allow it to actually work.

**Comment by bradfitz** [comment]
@normanr, I was both afraid (and excited) it might come to that. Glad to hear it should work! We have some WIP code to do that elsewhere, so that might give me motivation to finish it.

**Comment by DentonGentry** [comment]
Misc followups:

----

+ In strict mode where you tell Android a DNS service name like dns.google, it checks the validity of the certificate it gets back from the DNS server.
+ In opportunistic mode where Android just tries to make an SSL connection to port 853, it doesn't know a CN to check against so it accepts whatever it gets.

----

If we do make MagicDNS support TLS-over-DNS for keepsies (and I think we should) this probably involves generating a certificate per user domain, possibly via the LetsEncrypt API. Control might generate it to push down to clients, though I think we'd have a chicken-and-egg problem to work out. We'd need to either fetch the certificate before initiating the VPN connection (seems fine), or make sure the client won't need to make any more DNS lookups after it initiates the VPN connection before it can get the cert installed.

It may also be possible to generate one global certificate to be included in the APK, but as MagicDNS is implemented locally the private part of the cert would have to be included and would be the same for all users. That... doesn't seem great.

Doing DNS-over-TLS for a quick check could be done by generating a cert and installing the files on the local device, when we're ready to work on adding port 853+TLS support in tsdns.

----

I have not gotten the emulator to fail in opportunistic PrivateDNS mode, only in strict mode, but my phone experiences the opportunistic mode problem every time I leave the house and switch to T/Mobile's DNS servers via LTE. Unfortunately T/Mobile blocks access to those servers from outside their network, so I can't immediately make the emulator connect to them. I think I'd have to tether my Ubuntu system through the phone... which I'll probably do but is a PITA.

I'll also try to better characterize what it is about the T/Mobile DNS server which seems to trigger this worse for me than for most others who have reported it earlier in the bug. I've been assuming that the T/Mobile server is not recursively resolving on behalf of clients, but this is an assumption I haven't verified yet. I have a terminal app installed on my phone to investigate this. I might have to buy bifocals first. Good lord, the font.

**Comment by sreeram-boyapati** [comment]
I am facing this issue. I am trying to use DNS-over-TLS and TailScale on android. I want to use nextdns.io as it has some pre-configured blocklists

**Comment by DentonGentry** [comment]
Setting the tailscale DNS config to use nextdns.io means your DNS queries are being sent in cleartext, which might be fine depending on what you're trying to accomplish.

There is a workaround which allows PrivateDNS to be enabled on Android, albeit with significant compromises: on https://login.tailscale.com/admin/dns you need to both leave MagicDNS disabled and remove all Nameservers. This affects all of your Tailscale devices' DNS configuration, not just the Android device, but does allow the Android device to turn on PrivateDNS because it won't encounter the incompatibility with VpnServiceBuilder.addDnsServer().

**Comment by normanr** [comment]
btw, I added 8.8.0.0/16 to my advertised routes, and PrivateDNS (set to dns.google) now works when Tailscale is connected. So that confirms that PrivateDNS _can_ work, but it has to be routed to the VPN device on Android.

MagicDNS does not work, I because 100.100.100.100 is not used for lookups. I think the configured Private DNS server is looked up through Magic DNS - tcpdump shows it being looked up via the nameserver on the Tailscale DNS admin page.

When I switch to optimistic mode, then I see an attempt to connect to (the configured nameserver's) port 853, which I don't quite understand because Android should be trying 100.100.100.100, and I don't think the tailscale client can do DNS-over-TLS yet can it?

**Comment by DentonGentry** [comment]
I've been running Android 11 for some time, and mostly leave Private DNS set to Off at this point. I haven't run into this issue using that configuration.

We do have plans to implement DNS-over-TLS in Magic DNS a bit later this year, which should allow Private DNS to be set to Automatic and have it work.

**Comment by michaelclarkcuadrado** [comment]
This configuration had been completely broken in the latest android release. Not sure if I should open up a new issue. If you have a custom private dns server set, all networking on the device is broken. 

Last working android release is Tailscale 1.8.6-t28a8f9c90-gff16a75a65c. Disabling magicdns does not help.

**Comment by DentonGentry** [comment]
Please open a new issue.

Private DNS was previously working for you, with Tailscale supplying DNS servers? https://github.com/tailscale/tailscale/issues/915 implied that this configuration wasn't working for anyone.

**Comment by DentonGentry** [comment]
When Private DNS was working for you previously, was Magic DNS enabled?
Was Tailscale supplying the global DNS servers to use?

**Comment by michaelclarkcuadrado** [comment]
No, I don't use magicdns, it still affected this root issue with it disabled, but now all networking in sinkholed

**Comment by DentonGentry** [comment]
I had a vague hope that the work done to resolve https://github.com/tailscale/tailscale/issues/2102 would somehow fix this, but there is no change in behavior with Private DNS in Tailscale v1.10.

So far as I can tell the problem is an incompatibility within Android between the VpnService addDnsServer() and the PrivateDNS feature, see the comments from December about Strict Mode for details. Reinforcing this belief is that other Android VPN apps break with what look like similar symptoms:
- Cisco: https://quickview.cloudapps.cisco.com/quickview/bug/CSCvr47450
- Google Fi: https://www.reddit.com/r/ProjectFi/comments/ag273s/fi_vpn_breaks_googles_dnsovertls/

In earlier comments, @normanr mentions success by having the Private DNS IP route through the Tailscale network. I haven't tested this yet with 1.10.

**Comment by DentonGentry** [comment]
I built an image with https://github.com/tailscale/tailscale/pull/2435 but unfortunately it doesn't resolve this issue. With Android 12 if I set Private DNS to On in Settings with provider dns.google I get a System UI warning, "Network has no Internet access. Private DNS server cannot be accessed."

**Comment by bradfitz** [comment]
@DentonGentry, did you try with a Tailnet set to using DoH, or DoH w/ Google, or with an exit node?

I'm wondering what layer it's being blocked at.

**Comment by DentonGentry** [comment]
I tried it with DNS-over-TLS set to dns.google

**Comment by bradfitz** [comment]
I'm wondering what makes it think "Network has no Internet access", though.

**Comment by Xalaxis** [comment]
@bradfitz Unless I've misunderstood the wording, I always interpreted it as "Private DNS server cannot be accessed, therefore the network has no internet access (because we are blocking it)".

**Comment by bradfitz** [comment]
I guess that's one interpretation, but not how I read it. If that's the right interpretation, though, it still doesn't make sense, as we're not blocking access to `dns.google` so why wouldn't it work?

Unless Android is trying to resolve `dns.google` using our 100.100.100.100 and we're trying to resolve `dns.google` using plain ol' UDP:53 and failing.

But a) I'd be surprised if they didn't bake in 8.8.8.8 etc for `dns.google` into Android, and b) @DentonGentry tried a Tailscale Android build with the #2435 fix in, which means `100.100.100.100` would resolve `dns.google` using DNS-over-HTTP to 8.8.8.8, assuming @DentonGentry was using a Tailnet configured as such. That's what I was trying to ask above, but I'm not sure which Tailscale network Denton used and what its DNS settings were.

**Comment by bradfitz** [comment]
One quick test just for debugging would be to make 100.100.100.100 log all queries, to see what Android itself is querying.

And if it's really asking for `dns.google`'s records, we could bake in `8.8.8.8` etc.

**Comment by DentonGentry** [comment]
At the time I was using a personal Tailnet, which had MagicDNS enabled.

Android has no features for per-domain resolvers or Split DNS, so if Magic DNS is enabled then we tell Android to do all DNS resolution by sending to 100.100.100.100:53. We handle MagicDNS queries within Tailscale and forward the rest upstream.

However with PrivateDNS set to On, Android StrictMode blocks attempts to send to 100.100.100.100:53 so all DNS resolution is broken.

Android 12 has a more specific error that Private DNS is blocked, earlier Android versions said "Network has no Internet access"

**Comment by bradfitz** [comment]
Long shot, but: maybe Android permits 127.0.0.1:53 when PrivateDNS is on? Could we run our MagicDNS server on localhost instead? I'm sure it conservatively assumes that 100.100.100.100 is somewhere on the internet.

**Comment by normanr** [comment]
I can't remember in detail (and I can't test today, because I've just moved and my home connection isn't back up yet), but if I remember correctly, Android tried to connect to port 853 on the nameserver configured in tailscale (which for me is also my tailscale exit node). I just ran a port-redirect on that address using socat to redirect that to `dns.google:853` so that the connection ended up going where Android was expecting.

**Comment by bradfitz** [comment]
When I enable Private DNS on Android 11 (Pixel 3a), I see logcat output:

> 07-20 21:23:46.665  1539  2327 D DnsManager: sendDnsConfigurationForNetwork(101, [100.100.100.100], [bradfitz.gmail.com.beta.tailscale.net], 1800, 25, 8, 64, 0, 0, dns.google, [])

And then the menu item's gray subtitle says "Couldn't connect" and a notification says "Network has no internet access / Private DNS server cannot be accessed"

Will see if it's trying to connect to `100.100.100.100:853` next.

**Comment by bradfitz** [comment]
No, I don't see Android itself connecting to `100.100.100.100:853` over TCP.

But my debug logging does work. If I trigger TCP traffic there myself:

```
1|sargo:/ $ nc 100.100.100.100 853                                                                                                                                                                               
07-20 21:54:45.930 12362 12419 I com.tailscale.ipn: 7.9M/101.5M dns: XXX: magicdns traffic to TCP/853
^C
130|sargo:/ $ nc 100.100.100.100 1234                                                                                                                                                                            
07-20 21:56:37.822 12362 12489 I com.tailscale.ipn: 7.9M/101.5M dns: XXX: magicdns traffic to TCP/1234
^C
130|sargo:/ $ nc 100.100.100.100 52378                                                                                                                                                                           
07-20 21:56:43.347 12362 12439 I com.tailscale.ipn: 8.0M/101.5M dns: XXX: magicdns traffic to TCP/52378
^C
```

**Comment by bradfitz** [comment]
Ah, but I **do** see traffic to 100.100.100.100:853 when I use "Private DNS: Automatic" option:

```
07-20 22:13:25.140  1539  2327 D DnsManager: sendDnsConfigurationForNetwork(104, [100.100.100.100], [bradfitz.gmail.com.beta.tailscale.net], 1800, 25, 8, 64, 0, 0, , [100.100.100.100])
07-20 22:13:25.147 15086 15124 I com.tailscale.ipn: 10.4M/105.3M dns: XXX: packet out to TCP/100.100.100.100:853 from 100.122.37.18:37750
07-20 22:13:25.148 15086 15124 I com.tailscale.ipn: 10.5M/105.3M dns: XXX: magicdns traffic to TCP/853
07-20 22:13:26.164 15086 15124 I com.tailscale.ipn: 10.5M/105.3M dns: XXX: packet out to TCP/100.100.100.100:853 from 100.122.37.18:37750
07-20 22:13:26.166 15086 15124 I com.tailscale.ipn: 10.5M/105.3M dns: XXX: magicdns traffic to TCP/853
07-20 22:13:28.329 15086 15141 I com.tailscale.ipn: 10.6M/105.3M dns: XXX: packet out to TCP/100.100.100.100:853 from 100.122.37.18:37750
07-20 22:13:28.330 15086 15141 I com.tailscale.ipn: 10.6M/105.3M dns: XXX: magicdns traffic to TCP/853
07-20 22:13:32.377 15086 15201 I com.tailscale.ipn: 10.7M/105.3M dns: XXX: packet out to TCP/100.100.100.100:853 from 100.122.37.18:37750
07-20 22:13:32.378 15086 15201 I com.tailscale.ipn: 10.7M/105.3M dns: XXX: magicdns traffic to TCP/853
```

**Comment by bradfitz** [comment]
So I guess next step is to run a TCP server on 100.100.100.100:853 and see what SNI value (if any) Android's sending. If we're lucky, it's `dns.google` and we can proxy it. If it's unset, then we're out of luck for easy options, as I doubt Android would be happy with us answering DoT queries without a valid cert.

**Comment by apenwarr** [comment]
If simply proxying the session to dns.google somehow works, then I guess the client can't be checking the certificate? It doesn't know it's going to be reaching dns.google. I don't think that would be a very secure choice, but I don't know what they do instead.

**Comment by DentonGentry** [comment]
In Automatic mode (aka Opportunistic mode), Android does not check the certificate. It doesn't know the name to check against. Any DNS server it was going to send a packet to on UDP port 53, it YOLOs a TLS connection without any certificate validation instead and hopes for the best. At least the bits will be private on the wire.

Private DNS can also be turned "On" where you have to type in the name to check against like dns.google. Android will send those packets to 8.8.8.8, not 100.100.100.100.

**Comment by bradfitz** [comment]
The text at https://www.internetsociety.org/blog/2018/08/dns-privacy-in-new-android-9/ says that it checks the certificate chain validity, but doesn't check the name attached:

> Well there’s a setting in ‘Private DNS’ that says ‘Automatic’ which is pretty opportunistic, and after switching this on, the phone tries to send the DNS query to port 853 and falls back to port 53 if there’s no answer. Observing this through tcpdump confirmed the automatic setting does exactly what is expected – it figures out whether DNS-over-TLS is working on our recursive resolvers and starts using it. This option doesn’t validate server names or hashes, but it makes sure it can negotiate TLS 1.2 (and probably also validates the certificate chain that’s provided). However, in ‘Strict Mode’, the system does additionally validate the hostname where this is entered.

Which means we probably can't run a MagicDNS DoH server on-device to respond with MagicDNS results, as Android won't like our cert chain.

But proxying to `dns.google:853` would work fine, as it has a valid cert.

**Comment by normanr** [comment]
I think you _could_ run your own server for opportunistic mode. I don't remember if I tested with self-signed, private CA, or LetsEncrypt certs. If is only works with the latter, then #1235 could be used to get a valid cert.

For strict mode, it would need to be tunneled to the correct upstream server. Note that in strict mode the local DNS resolver (which would be 100.100.100.100), is used to lookup the IP address(es) of the configured authentication domain name. That may be a bit too late to ensure that the resolved address is always routed over tailscale, but may give tailscale an opportunity to sniff the target hostname if there's no SNI in the TLS hello.

refs:
- https://developers.google.com/speed/public-dns/docs/dns-over-tls
- https://developers.google.com/speed/public-dns/docs/using#android
- https://android-developers.googleblog.com/2018/04/dns-over-tls-support-in-android-p.html
- https://datatracker.ietf.org/doc/html/rfc7858

**Comment by dr8** [comment]
An important use case to consider is self-hosted Private DNS (for adblocking, etc.). This would be with Private DNS set to the third option with a not-well-known DNS provider hostname provided.
In this case (on Android 12 beta), you get an error 'Private DNS server could not be reached'; presumably because the initial DNS lookup fails to resolve the address for the Private DNS server itself. Obviously, this use case wouldn't be fixed by hard-coding addresses for various well known services, like 8.8.8.8.

**Comment by bradfitz** [comment]
@dr8, for that case you can just use Tailscale, though.

**Comment by dr8** [comment]
Sure, that can work, but requires Tailscale to be always on. I want to activate Tailscale based on need to access certain machines, to reduce battery drain. As it stands, Private DNS settings have to be manually flipped in system settings every time you stop and start Tailscale.

**Comment by bradfitz** [comment]
@dr8, unfortunately there doesn't seem to be anything we can do to make Private DNS with a specific hostname work when Tailscale's both on & off. But we can do something about Private DNS in "Automatic" mode when Tailscale's on, so I'm focusing on that first.

If anybody has clever ideas about how we can coexist with an explicit Private DNS setting, I'm all ears.

**Comment by dr8** [comment]
@bradfitz thanks for that. I appreciate what you're trying to do. 

For reference, in WireGuard (and I presume other VPN clients), DNS lookups to the WireGuard-provided DNS server are limited to looking up the PrivateDNS server, and all other DNS look-ups continue to go to the PrivateDNS server as normal (albeit through the tunnel first).

A suggestion: given that with MagicDNS enabled and PrivateDNS is set to a specific hostname all connectivity is lost, wouldn't it be better to selectively disable MagicDNS on those devices? (i.e. allow the device to continue to use PrivateDNS and accept loss of MagicDNS benefits). I'd rather lose MagicDNS than lose all internet connectivity.

Of course, for folks like me who are self-hosting their own PrivateDNS server, there's nothing to stop us forwarding the relevant queries to the tailscale DNS server, although that does make it a bit more complicated.

**Comment by DentonGentry** [comment]
Since this bug was filed, Split-DNS support has been added https://tailscale.com/kb/1054/dns/

On platforms where the OS native resolver supports per-domain DNS servers, like Windows, the tailscale client will push the configuration into the OS to handle.

Android's resolver has no such functionality. On Android, we set 100.100.100.100 as the primary resolver and have it forward requests for the split-DNS domain to one set of servers and everything else to the configured fallback servers.

So yes, Split DNS configuration can result in hitting this issue.

**Comment by belimawr** [comment]
I'm also having similar issues over LTE, when Tailscale MagicDNS is enabled, my Android (Google Pixel 3) cannot resolve some hosts. I haven't managed to identify the hosts, but Twitter images don't load, and some other apps won't work properly (like [Berlin's public transportation app](https://play.google.com/store/apps/details?id=de.eos.uptrade.android.fahrinfo.berlin).

Disabling Tailscale or MagicDNS solves the problem for me.

I have had no issues on my Linux machines.

It has been like that since months ago, maybe by the time this issue was opened.

I'm happy to help reproducing/debugging it. My Android development knowledge is limited, but I'm happy to follow any instruction/guides you folks point me at.

**Comment by DentonGentry** [comment]
Do you have "Private DNS" enabled in the Android settings, to either Automatic or On? If so, then you might be hitting this issue. If not, you're having some other DNS-related issue on Android.

The main thing would be, shortly after noting a DNS problem, to send a bugreport to support@tailscale.com or file a new issue with it. The Android app added a "bug report" item in its menu in the 1.20 release (I think, though maybe 1.22).

**Comment by belimawr** [comment]
> Do you have "Private DNS" enabled in the Android settings, to either Automatic or On? If so, then you might be hitting this issue. If not, you're having some other DNS-related issue on Android.

Yes, Private DNS is set to Automatic.

> The main thing would be, shortly after noting a DNS problem, to send a bugreport to [support@tailscale.com](mailto:support@tailscale.com) or file a new issue with it. The Android app added a "bug report" item in its menu in the 1.20 release (I think, though maybe 1.22).

I'll keep that in mind!

**Comment by DentonGentry** [comment]
https://github.com/tailscale/tailscale/pull/4598 moved the MagicDNS feature from the bespoke packet handling implementation it has used until this point over to instead be a [netstack](https://pkg.go.dev/inet.af/netstack) listener. It also implemented DNS-over-TCP support, where MagicDNS previously had only handled UDP. These were both necessary preconditions to be able to implement DNS-over-TLS support for PrivateDNS on Android.

**Comment by ignoramous** [comment]
> So far as I can tell the problem is an incompatibility within Android between the VpnService addDnsServer() and the PrivateDNS feature, see the comments from December about Strict Mode for details. 

An app I co-develop does use `VpnService#addDnsServer` API. Private DNS (if user sets it) does work without hindrances. Android sends A/AAAA queries to VPN's DNS endpoint for resolution of the Private DNS domain name. May be, queries sent to quad100 are routed to another link (which refuses to answer those DNS queries) and not to tailscale (though, that's unlikely given vpn routes are higher priority on Android?).

> Which means we probably can't run a MagicDNS DoH server on-device to respond with MagicDNS results, as Android won't like our cert chain.

Do note that, due the recent DST -> ISRG root change, [Android's DoT impl with Let's Encrypt certs](https://blog.jeroenhd.nl/article/let-s-encrypt-dns-over-tls-and-android) doesn't work whatsoever.

**Comment by DentonGentry** [comment]
> An app I co-develop does use VpnService#addDnsServer API. Private DNS (if user sets it) does work without hindrances. Android sends A/AAAA queries to VPN's DNS endpoint for resolution of the Private DNS domain name.

Well **that** is fascinating, from what we could see in https://github.com/tailscale/tailscale/issues/915#issuecomment-751706094 it was a StrictMode policy preventing DNS packets from being sent to UDP port 53. The error references `android.os.strictmode.UntaggedSocketViolation: Untagged socket detected`, do you call any kind of API about tagging your DNS listener in some way?

**Comment by ignoramous** [comment]
We do not tag any sockets in kotlin/golang; the outgoing doh/dnsrypt/dns53 connections the app makes to resolve private-dns domain names are the usual connections it would otherwise make in servicing any other DNS queries.

I see strict-mode related logs too, but it proceeds with `OK` in our case.

Like tailscale, the app handles dns udp/tcp socksified by netstack, with the only difference I see is [this netns impl for Android](https://github.com/tailscale/tailscale/blob/758c37/net/netns/netns_android.go) ([is this jni.Do attached to ui-thread?](https://github.com/tailscale/tailscale-android/blob/84b484a9541d0549ce7073f03f26264a14801f9c/cmd/tailscale/main.go#L427)).

As a stop-gap, [untagged sock penalty can be disabled](https://stackoverflow.com/questions/54212614), even if it isn't pretty...

**Comment by Chaz6** [comment]
Is there any progress on this? I cannot use Tailscale on my Android device due to this bug. I have version 1.32.0.

**Comment by DentonGentry** [comment]
PrivateDNS in Automatic mode should work today.

PrivateDNS in On mode, where it requires a proper certificate and sets strict mode to prevent UDP DNS packets from working, is stuck at the difficulty of provisioning a certificate. 100.100.100.100 is not a DNS server on the Internet, it is a tiny DNS server running as part of the tailscale app: https://tailscale.com/blog/magicdns-why-name/

To provision a certificate for quad-100 we need to either:
- get each phone to provision a [certificate using LetsEncrypt](https://tailscale.com/blog/tls-certs/). This is possible but has the downside that the existence of certificates is recorded on a public ledger, which exposes more information about Android use within an organization than appears reasonable.
- issue just one certificate for quad-100, and push the private certificate down to the app. This would be immediately extracted, which is bad.

**Comment by normanr** [comment]
Making the private key available will result in the certificate being rapidly revoked (as required by the CA/Browser forum baseline requirements), as sslip.io found out the hard way (https://github.com/cunnie/sslip.io/issues/6#issuecomment-765421501).

You probably have to issue a per-device certificate (using the current let's encrypt integration), but that would mean that each device would need to use a different value for their upstream resolver. In terms of privacy, you _could_ issue all quad-100 certs from a common domain, i.e. instead of quad-100.yaq-bebop.ts.net, it would be random-id.quad-100.ts.net. That way public information about Android usage would be aggregated across all of Tailscale and not tied to any specific organisation/tailnet.

**Comment by bradfitz** [comment]
> Making the private key available will result in the certificate being rapidly revoked (as required by the CA/Browser forum baseline requirements), as sslip.io found out the hard way (https://github.com/cunnie/sslip.io/issues/6#issuecomment-765421501).

You probably have to issue a per-device certificate (using the current let's encrypt integration), but that would mean that each device would need to use a different value for their upstream resolver. In terms of privacy, you could issue all quad-100 certs from a common domain, i.e. instead of quad-100.yaq-bebop.ts.net, it would be random-id.quad-100.ts.net. That way public information about Android usage would be aggregated across all of Tailscale and not tied to any specific organisation/tailnet.

Yeah, now that we've got some of the internal naming issues cleaned up from the Magic DNS GA launch, that's sounding like the best option here.

But rather than `*.quad-100.ts.net` we might just do `<randomhex>.node.ts.net` and not make it necessarily just for quad 100. I can think of at least one other thing we could use that for too.

/cc @DentonGentry @maisem @crawshaw

**Comment by DentonGentry** [comment]
We'll need to make it easy to copy `<randomhex>.node.ts.net` to paste into the Android Settings. For strict Private DNS mode you have to enter the name of the DNS server so it can check certificates.

**Comment by bradfitz** [comment]
> We'll need to make it easy to copy <randomhex>.node.ts.net to paste into the Android Settings. 

No Android API to set that, eh? 😢

**Comment by ignoramous** [comment]
> No Android API to set that, eh? cry

Only device-poilcy-managers: [DevicePolicyManager#setGlobalPrivateDnsModeSpecifiedHost](https://developer.android.com/reference/android/app/admin/DevicePolicyManager#setGlobalPrivateDnsModeSpecifiedHost(android.content.ComponentName,%20java.lang.String)). Though, some bits are [deprecated in Android 10+](https://developers.google.com/android/work/device-admin-deprecation), but it shouldn't matter as tailscale would [enroll into the EMM program](https://androidenterprisepartners.withgoogle.com/emm/), eventually?

**Comment by normanr** [comment]
yea, but also:
> In case a VPN is used in conjunction with Private DNS resolver, the Private DNS resolver must be reachable both from within and outside the VPN. Otherwise, the device may lose the ability to resolve hostnames as system traffic to the resolver may not go through the VPN.

If the user sets Private DNS with `<randomhex>.node.ts.net`, then when Tailscale is offline, the device won't have DNS -- and won't be able to reconnect to Tailscale, because there's no DNS.

I understand why VPN shouldn't be able to override Private DNS (with Host) for security reasons, but it would be nice to have support for split DNS (VPN would declare which sub-domains are allowed, user would be prompted to allow, and choice could be remembered?)

I guess for Android _today_ the best Tailscale could do is just make it obvious to the user (alert them) if Private DNS (with Host) is currently configured, say that it's incompatible, and that they need to reconfigure it to Private DNS (Automatic).

**Comment by MatMaul** [comment]
Would it be possible to have a way to disable MagicDNS, either in the Android app or in the management interface per device ?
This is quite annoying, and while not having MagicDNS for technical reason is fine for me, needing to disable my private DNS each time I want to use Tailscale on my Android is highly annoying.

Thanks !

**Comment by Alpaca131** [comment]
Hi! 
I also experienced this issue after enabling MagicDNS.
I managed to solve this issue by using the automation app(e.g. Automate) to edit system setting variable "private_dns_mode" in Global table, and I would like to make a suggestion to make a function to edit this automatically when VPN connects and disconnects.

This requires users to grant permissions "WRITE_SECURE_SETTINGS" to the app via ADB, so maybe the dev team doesn't want to do this, though I think this is a possible way to solve this problem because only users who want to use "auto private-dns control" need to use ADB.

**Comment by xdevs23** [comment]
I think we should rather focus on getting support for this into AOSP itself because then we would be able to fix this without dirty hacks or workarounds.

At least an integration into VpnService to set private dns servers would be a good starting point. The user would probably need to approve this by confirming a system dialog, but other than that I think it's reasonable.

**Comment by mikklfr** [comment]
Thanks @Alpaca131 I managed to get something working using Automate and published a community flow that watches the Android app logs for Tailscale connection.
(The task for VPN connection surveillance was not working on my phone).

https://llamalab.com/automate/community/flows/44164 
Hope it can help others a bit since that issues is still a bit painful on a day to day basis.

**Comment by Alpaca131** [comment]
> I think we should rather focus on getting support for this into AOSP itself because then we would be able to fix this without dirty hacks or workarounds.
> 
> At least an integration into VpnService to set private dns servers would be a good starting point. The user would probably need to approve this by confirming a system dialog, but other than that I think it's reasonable.

I don't think this will work soon, because this needs an additional API for managing Private-DNS settings. 
Currently Android doesn't have such an API, and I suppose such a change will be made after Android 15, which takes a long time. Android 14 beta will begin from 2023/4 and I don't think Google will add another feature at this time to Android 14.

**Comment by Alpaca131** [comment]
> Agreed on the fact that AOSP itself should support it, meanwhile Automate actually helped me a lot.
> 
> Thanks @Alpaca131, I managed to get something working using Automate and published a community flow that watches the Android app logs for Tailscale connection. (The task for VPN connection surveillance was not working on my phone).
> 
> https://llamalab.com/automate/community/flows/44164 Hope it can help others a bit since that issues is still a bit painful on a day to day basis.

Thanks! What I made was using notification logs, since the connection change trigger didn't work, it seems your solution is smarter. 
However, is the log trigger `.*.tun0 is deactivated` right? In my device, the trigger of VPN deactivation is not working.

**Comment by mikklfr** [comment]
It works on my side but it's a Pixel 6 Pro, maybe the interface name differs from devices.
I also use launcher flow shortcuts, it allows to trigger the switch using only one click on the launcher and avoids to read the logs and therefore maybe save a bit of battery life.

Enable
https://llamalab.com/automate/community/flows/44195
Disable
https://llamalab.com/automate/community/flows/44194

**Comment by xdevs23** [comment]
> Currently Android doesn't have such an API, and I suppose such a change will be made after Android 15, which takes a long time. Android 14 beta will begin from 2023/4 so I don't think Google will add another feature at this time to Android 14.

@Alpaca131 I believe that it isn't unlikely that Google would add such a feature quickly. Sometimes they change a LOT even just between two revisions, causing major merge conflicts. While they won't introduce a public API into the code inbetween revisions, it is possible that they add something like this before the next major release if it is deemed important.

Regardless of Google's timeline, I would also be okay with adding such a patch to custom ROMs until Google adds the API. I personally don't want to go through the hoops of submitting such a CL to Google, so if anyone here is willing to do that, it would be much appreciated. I would then pick this into my custom ROM, and if Tailscale decides to support this unofficial API until Google makes it public, this issue could be fixed for many people already.

**Comment by normanr** [comment]
Part of the long timeline with Android changes is that anything that changes the UI needs translation, which significantly increases the time it takes to get it released. Also it looks like this is being reported in [various](https://issuetracker.google.com/issues/119878172) [ways](https://issuetracker.google.com/issues/141674015), with no plans to fix it. So I wouldn't expect any fixes from AOSP any time soon.

Something I just realized (and sorry if I'm late to the party and this is what DentonGentry and bradfitz have been thinking about all the time), is that not only could Tailscale could serve `<randomhex>.node.ts.net` from `100.100.100.100` when you're connected to Tailscale (and then serve it locally from tailscaled), but _also_ with a public IP address when you're _not_ connected to Tailscale (from the similar DNS servers that are used for the LetsEncrypt integration). The `randomhex` could be per-node, or per-network, and the admin panel would allow you to select which 'upstream' resolver you want to use (or maybe tailscale would do the resolving?!) It does mean trusting Tailscale with your DNS queries even when Tailscale is not enabled, and Tailscale has to be willing to serve those queries too. Tihis is very similar to what services like NextDNS are doing.

Also if you really really want to use a single certificate for all devices, you don't _have_ to push the private key for the shared certificate to the clients. You could host a keyserver and use Keyless SSL (as Cloudflare calls it). Then the client would use it's connection back to the tailscale co-ordination server to exchange the data needed to establish the TLS connections.

**Comment by xdevs23** [comment]
> Part of the long timeline with Android changes is that anything that changes the UI needs translation, which significantly increases the time it takes to get it released. Also it looks like this is being reported in [various](https://issuetracker.google.com/issues/119878172) [w](https://issuetracker.google.com/issues/141674015)

@normanr That's a really good point, I did not think about that. I still would accept a patch that I can pick into my custom ROM so that this issue is fixed until Google does something about it.

**Comment by Alpaca131** [comment]
I have a new proposal about this issue.
Is it possible to enable/disable Magic DNS per device?
I think this is the most simple, and clean way to solve this.
At least, for me, it's totally fine if I can't use Magic DNS on my smartphone, because I see that function mostly on PC or Chromebook.

**Comment by dr8** [comment]
> I have a new proposal about this issue. Is it possible to make a toggle that enable/disable Magic DNS per device? (On a client side) I think this is the most simple, and clean way to solve this. At least, for me, it's totally fine if I can't use Magic DNS on my smartphone, because I use that function mostly on PC or Chromebook.

I agree, and I suggested [the same back in July](https://github.com/tailscale/tailscale/issues/915#issuecomment-886018024) 
Surely opting out of magic DNS is better than losing all connectivity?

**Comment by Alpaca131** [comment]
> > I have a new proposal about this issue. Is it possible to make a toggle that enable/disable Magic DNS per device? (On a client side) I think this is the most simple, and clean way to solve this. At least, for me, it's totally fine if I can't use Magic DNS on my smartphone, because I use that function mostly on PC or Chromebook.
> 
> I agree, and I suggested [the same back in July](https://github.com/tailscale/tailscale/issues/915#issuecomment-886018024) 
> Surely opting out of magic DNS is better than losing all connectivity?

Yes, certainly.
I found the same suggestions here, including yours, so it turned out that it wasn't a new idea at all. 😂

**Comment by Alpaca131** [comment]
@DentonGentry 
Do you have any schedule or intent to implement this function?

**Comment by DentonGentry** [comment]
I understand that this is a problem, and we do intend to fix it.
When there is progress made it will be posted here, we do not currently have a schedule to expect.

**Comment by Alpaca131** [comment]
> I understand that this is a problem, and we do intend to fix it. When there is progress made it will be posted here, we do not currently have a schedule to expect.

Thank you for answering! I'm glad to hear that.
Tailscale is the core of my development and entertainment (like Minecraft), and I really appreciate your daily development.

**Comment by yegle** [comment]
Hmm, would it be possible to add a toggle to the Android app to disable MagicDNS, until this bug is fixed?

It looks like the disable toggle is available in the `tailscale` CLI but not on Android app.

**Comment by patriiiiiiiiiick** [comment]
Do we indeed work around this issue by disabling MagicDNS, regardless of other settings? I am trying that at the moment.

Pixel 6, Android 13

**Comment by Alpaca131** [comment]
Or, maybe there should be a notice like this to the user. 
(Screenshot from Tunnelbear, Pixel 7, Android 14)
![Screenshot_20230824-202220](https://github.com/tailscale/tailscale/assets/66709717/9ea7045f-9d86-441a-bd91-ca53ebafd464)

**Comment by nullbyto** [comment]
A better solution for the workaround recommended above, is to add the option to disable using DNS only for android device instead of disabling [magicDNS](https://tailscale.com/kb/1081/magicdns/) for all devices, for Windows and MacOS this option is available. It would be good to add this to the app on android, and/or in addition an option from the admin panel to disable a particular device from using DNS.

**Comment by aladh** [comment]
This issue has been open for almost three years. It would be great if the proposed solution of adding a MagicDNS toggle in the Android app could be prioritized.

**Comment by WreckingU** [comment]
Yeah this is a deal breaker for me. MagicDNS breaks a lot of functionality when switching wifi to cellular.

**Comment by masx200** [comment]
https://github.com/tailscale/tailscale/issues/7311

**Comment by mackncheesiest** [comment]
https://xkcd.com/2881/ made me think of this thread

Much <3 to the tailscale devs regardless :)

I've found a slight solution where "when using mullvad exit nodes on tailscale, using mullvad's resolver in the private dns settings (i.e. base.dns.mullvad.net) continues to work fine".

**Comment by TheWonderfulTartiflette** [comment]
> This issue has been open for almost three years. It would be great if the proposed solution of adding a MagicDNS toggle in the Android app could be prioritized 🙂

Yep, can confirm it's happening to me as well, same issue as everyone else here

**Comment by tmplt** [comment]
I think this issue can be side-stepped by putting the Andoid device on a separate tailnet where Magic DNC is disabled, and wanted services are [shared](https://tailscale.com/kb/1084/sharing). I have yet to verify. Does this require a wholly separate account or can we be allotted more tailnets?

**Comment by unixfox** [comment]
I'm not quite sure how to reflect that for tailscale users but with [headscale](https://github.com/juanfont/headscale), if you control the private dns that you want to use and have your own domain name then you can solve this issue.

So basically my [`base_domain`](https://github.com/juanfont/headscale/blob/main/config-example.yaml#L269C3-L269C14) is a real domain name. Every device registered on my tailscale network is under `tailscale.mydomain.com`, example for the device named `android` it's `android.tailscale.mydomain.com`.

I have hosted my own private dns (DNS-Over-TLS) server running with [AdGuardHome](https://github.com/AdguardTeam/AdGuardHome) under the domain name: `doh.tailscale.mydomain.com`. When I'm not using tailscale, the traffic is sent to the public IP of that server because it's registered as a DNS record on my domain.

What I did is to name in my tailnet my server `doh`, this way when I enable tailscale, the traffic is sent through the IP address of the server in the tailnet instead of the public IP address. Somehow this works fine!

<img src="https://github.com/tailscale/tailscale/assets/4016501/298d7308-3657-44e1-9a94-a977674c9df7" width="300" />

**Comment by agottardo** [comment]
Reproduced on 1.65.0 with a Google Pixel 7a running Android 14. When the Private DNS setting is set to a custom DoT server and Tailscale is running, no DNS queries can go through in Chrome or other apps.
```
19.9M/323.9M wgengine: Reconfig: configuring DNS
19.9M/323.9M dns: Set: {DefaultResolvers:[] Routes:{tailnet-name.ts.net.:[] ts.net.:[199.247.155.53 2620:111:8007::53]}+65arpa SearchDomains:[tailnet-name.ts.net.] Hosts:445}
19.9M/323.9M dns: Resolvercfg: {Routes:{.:[192.168.1.1] ts.net.:[199.247.155.53 2620:111:8007::53]} Hosts:445 LocalDomains:[tailnet-name.ts.net.]+65arpa}
19.9M/323.9M dns: OScfg: {Nameservers:[100.100.100.100] SearchDomains:[tailnet-name.ts.net. home.] }
```

Resolvercfg indicates `192.168.1.1` is being used for all queries which is the local network forwarder as expected. A system notification appears, indicating the Private DNS server cannot be reached anymore.

**Comment by chennin** [comment]
Tailscale 1.66.x on Android was [rebuilt](https://tailscale.com/blog/android) and includes a "Tailscale DNS" toggle, which is one of the workarounds asked for here. (Sign up for the beta in the play store if you want the update sooner). Thank you for putting that option in.

Quoting the blog:

> We’ve fixed some of the most reported issues including DNS resolution

Does that refer to *this* issue? Could someone from Tailscale please comment on whether the new app is intended to fix the underlying bug here?

I've been using it for a couple days regularly, not specifically attempting to replicate, and I haven't lost DNS *yet*.

**Comment by shawnz** [comment]
> Tailscale 1.66.x on Android was [rebuilt](https://tailscale.com/blog/android) and includes a "Tailscale DNS" toggle, which is one of the workarounds asked for here.

Turning off that toggle successfully works around the issue for me as well, and turning it on continues to give the old broken behaviour as expected.

**Comment by yegle** [comment]
I'm still seeing the same issue with the new DNS option disabled.

![Screenshot_20240516-115929.png](https://github.com/tailscale/tailscale/assets/146133/fb24da18-df2b-4ecf-ac1c-0b6e1e121320)

![Screenshot_20240516-115946.png](https://github.com/tailscale/tailscale/assets/146133/380ec694-d1e8-4290-8b62-68d18a3549b0)

(I also tried `dns.google` as the DoT server to no avail).

**Comment by tmplt** [comment]
Much better GrapheneOS compatibility at a first glance: MagicDNS works and Internet access remains.

**Comment by agottardo** [comment]
I can confirm no specific work has been done in v1.66 for this issue, but it is on our radar... I did briefly try to identify the reason why it's broken but it's something more systemic in our DNS setup. Glad the workaround is helping in the meantime.

cc @kari-ts

**Comment by BJReplay** [comment]
> I'm still seeing the same issue with the new DNS option disabled.
> 
I'm seeing the similar issues:
- If I leave the toggle on (Use Tailscale DNS), Private DNS is completely broken.
- If I turn the toggle off, Private DNS works until I have an upstream network change (e.g. WiFi to LTE or vice versa) and then it breaks until I turn off Tailscale, turn off Private DNS, turn on Private DNS, and turn on Tailscale.

**Comment by nugget210** [comment]
I setup Tailscale for the first time yesterday and it's working perfectly on all of my devices except for my Android.  I'm getting exactly the same issues described above.  I'm assuming this is still being looked at?

**Comment by BJReplay** [comment]
>  I'm assuming this is still being looked at?

I'm not convinced.

I raised an issue, and got a response a couple of weeks later (which is fine - the team is busy, and I am a free customer) indicating that the Android error message is a false positive, and that 30 seconds later, all will be OK under the covers.

I have followed up indicating that I know that it is not because I rely on DNS rewrites by my private DNS to reach some internal sites with a FQDN, and they are a) unreachable, and b) do not resolve, indicating that the private DNS is, as Android reports, unreachable.

**Comment by agottardo** [comment]
We know this is important to you and want to reassure you this is high on the priority list. We spent the last couple weeks working on long-standing feature requests like ping and [app-based split tunneling](https://github.com/tailscale/tailscale-android/pull/435), shipping in the next app update. This is next on the list of things to tackle.

Unfortunately, it's a hard problem to diagnose and the immediate findings have been initially inconclusive. When Private DNS is enabled, no queries seem to reach Tailscale so we can't isolate this to an issue on our side, and on the other hand Android DNS logging does not provide any insights into what is causing the issue beyond telling us that DNS queries are failing.

**Comment by BJReplay** [comment]
> We know this is important to you and want to reassure you this is high on the priority list.

Thanks @agottardo! Your reply is appreciated.

Your support team for in-app support requests seem less convinced there is a problem😅

When you return to this issue, if logcats will help, please ask and I will be pleased to supply.

**Comment by kari-ts** [comment]
I was able to repro. Interestingly while DNS is not working for me, my logs do show TLS servers: 

`updatePrivateDns(120, PrivateDnsConfig{strict:dns.google/[/2001:4860:4860::8844, /2001:4860:4860::8888, /8.8.4.4, /8.8.8.8], dohName=, dohIps=[], dohPath=, dohPort=-1})
sendDnsConfigurationForNetwork(121, [2001:558:feed::2, 2001:558:feed::1, 75.75.75.75, 75.75.76.76], [hsd1.ca.comcast.net], 1800, 25, 8, 64, 0, 0, dns.google, [2001:4860:4860::8888, 2001:4860:4860::8844, 8.8.4.4, 8.8.8.8], [1], false)
`

I'll keep digging into this.

**Comment by kari-ts** [comment]
I was able to repro. Interestingly while DNS is not working for me, my logs do show TLS servers:

`updatePrivateDns(120, PrivateDnsConfig{strict:dns.google/[/2001:4860:4860::8844, /2001:4860:4860::8888, /8.8.4.4, /8.8.8.8], dohName=, dohIps=[], dohPath=, dohPort=-1})

sendDnsConfigurationForNetwork(121, [2001:558:feed::2, 2001:558:feed::1, 75.75.75.75, 75.75.76.76], [hsd1.ca.comcast.net], 1800, 25, 8, 64, 0, 0, dns.google, [2001:4860:4860::8888, 2001:4860:4860::8844, 8.8.4.4, 8.8.8.8], [1], false)
`

I'll keep digging into this.

**Comment by kari-ts** [comment]
I was able to repro. Interestingly while DNS is not working for me, my logs do show TLS servers:

```
updatePrivateDns(120, PrivateDnsConfig{strict:dns.google/[/2001:4860:4860::8844, /2001:4860:4860::8888, /8.8.4.4, /8.8.8.8], dohName=, dohIps=[], dohPath=, dohPort=-1})

sendDnsConfigurationForNetwork(121, [2001:558:feed::2, 2001:558:feed::1, 75.75.75.75, 75.75.76.76], [hsd1.ca.comcast.net], 1800, 25, 8, 64, 0, 0, dns.google, [2001:4860:4860::8888, 2001:4860:4860::8844, 8.8.4.4, 8.8.8.8], [1], false)
```

I'll keep digging into this.

**Comment by JasonHK** [comment]
> > I'm still seeing the same issue with the new DNS option disabled.
> > 
> I'm seeing the similar issues:
> - If I leave the toggle on (Use Tailscale DNS), Private DNS is completely broken.
> - If I turn the toggle off, Private DNS works until I have an upstream network change (e.g. WiFi to LTE or vice versa) and then it breaks until I turn off Tailscale, turn off Private DNS, turn on Private DNS, and turn on Tailscale.
> 

Interestingly, I’ve only encountered the first behaviour but not the second one.

**Comment by kari-ts** [comment]
I was able to repro. 

```
2024-07-23 12:14:28.482  1499-2066  DnsManager              system_server                        W  updatePrivateDns(182, PrivateDnsConfig{strict:dns.google/[/2001:4860:4860::8888, /2001:4860:4860::8844, /8.8.4.4, /8.8.8.8], dohName=, dohIps=[], dohPath=, dohPort=-1})
2024-07-23 12:14:28.482  1499-2066  ConnectivityService     system_server                        D  Setting DNS servers for network 182 to [/2001:558:feed::2, /2001:558:feed::1, /127.0.0.1]
```

**Comment by raggi** [comment]
@kari-ts @agottardo and I looked at this more today, and now understand the problem better.

Android Private DNS behavior currently is that when private dns is set, Android ignores any DNS configuration that we set via the VPN APIs.

It recognizes the VPN interface as a new interface, and performs an unconditional interface binding on the DNS client sockets, sending all DNS client traffic to the VPN interface.

If you have an exit node configured, this will be fine, but if you do not, then there is no route on the VPN interface matching the traffic being sent from the DNS client socket, and the packets are dropped by the kernel in routing decision.

The choice Android is making here to ignore DNS configuration set via the VPN API is a complicated choice, and it would be better if there were more APIs and UI/UX covering this behavior, see remaining issue further down.

The choice Android is making to ignore routing and send packets to somewhere that can't possibly route them is essentially a platform bug, this behavior is clearly broken. It is essentially assuming that all VPNs are global routers, which is not the case.

We are now looking in to mitigations where we locally add routes matching the upstream DNS servers configured in private DNS so we can receive the packets, and forward them on via an appropriate interface.

There is a remaining issue which is that because of the behavior where private DNS ignores the configuration that we are setting via the configuration APIs, MagicDNS can not work while private DNS is enabled, as we are prevented from answering any of the necessary DNS queries. If we are able to implement a routing mitigation, we should add a prominent health check warning explaining that MagicDNS can't work while PrivateDNS is configured.

**Comment by luigir-it** [comment]
> private DNS ignores the configuration that we are setting via the configuration APIs

Does that mean that private DNS and MagicDNS won't work together even with an exit node set up?

**Comment by agottardo** [comment]
> Does that mean that private DNS and MagicDNS won't work together even with an exit node set up?

@luigir-it Correct. Android sends all queries to the DoT server you specified in the Android network settings, so our resolver doesn't get a chance to provide answers for *.ts.net hostnames.

**Comment by BJReplay** [comment]
> @kari-ts @agottardo and I looked at this more today, and now understand the problem better.

@raggi that is excellent analysis, and you have clearly investigated this problem in intricate detail.

My manual work around (which has been easy enough, for me, so far) has been to manually add DNS entries or re-writes to both my Private DNS and Tailscale so it doesn't matter which is providing resolution, both resolve internal names the same way.  With only a few devices in a home network environment and a few cloud devices via Tailscale, it has been a relatively low maintenance burden - but still one that I manage to forget each time I add a new device, until I realise why I can't connect.

**Comment by raggi** [comment]
@BJReplay makes sense.

If you're able to leave tailscale on at all times on the phone then a fully working and sound configuration is to set "override dns" in the admin panel to a secure dns provider, and then disable private dns in android. We'll then be handling all DNS all the time in the client, and do DoH to that provider, and magicdns will work normally.

This does change some power accounting, for example tailscale client time spent servicing dns now becomes power accounting for tailscale rather than system, but otherwise should be essentially the same as the system doing the work.

Hopefully we can eventually persuade android to provide sufficient API surface that both features can be enabled at the same time without the need for extra workarounds, for now at least we should soon be able to workaround the larger android bug here so "the internet works" while privatedns is configured and tailscale is on.

**Comment by BJReplay** [comment]
> is to set "override dns" in the admin panel to a secure dns provider

The issue with this at the moment is that you can choose a pre-defined one from a list or add an IP.

You can't (yet) specify a custom (or choose more than two DoH pre-defined) name-servers, and, specific to this particular bug, you can't specify a DNS-over-TLS nameserver in this dialog.

![image](https://github.com/user-attachments/assets/2ee69846-b2a9-4eb9-9a96-b7bcfc92b883)
![image](https://github.com/user-attachments/assets/d0de8470-a59d-4634-ac89-fdaa63adc0a9)


So, to the extent that users are happy using those, sure 😉

But I'm not, which is why I'm running my own in the first place.

But it is a fair point, and not an unreasonable suggestion, given the situation you find yourself in, and the decisions made withing android to assume that all VPNs are global routers which was probably always true back when... 

android phones were not connecting to VPNs.

**Comment by kari-ts** [comment]
Closing this issue for now. Please see [https://issuetracker.google.com/issues/141674015#comment10](https://issuetracker.google.com/issues/141674015#comment10) for the Android ticket. 

With the Android private DNS enabled, Tailscale is unable to disambiguate between DNS and other traffic - we aren't able to implement apriori forwarding, and without being able to disambiguate traffic from the first packet, we can't reliably know if a given connection should be proxied or treated as normal network traffic. 

Future improvements include health warnings flagging that Tailscale and private DNS are not interoperable.

**Comment by ignoramous** [comment]
Clickable link: https://issuetracker.google.com/issues/141674015#comment10

**Comment by eiqnepm** [comment]
I have Use Tailscale DNS disabled in Android but private DNS is still not working when Tailscale is connected

**Comment by lorengordon** [comment]
Encountered this issue as I was trying to use NextDNS with multiple profiles to configure different settings for different devices. Tailscale only allows setting one global NextDNS server, rather than one per device. I setup Private DNS on the phone with the new NextDNS profile, and that worked fine. Disabled "Use Tailscale DNS" in the tailscale app settings, then connected to the Tailscale VPN, but then Private DNS breaks as covered by this issue.

At least for my use case, an alternative solution would be per-device DNS Settings in the Tailscale Admin Console...

**Comment by Atemu** [comment]
This is a long thread and I haven't read every message, so apologies if this has already been mentioned before:

Would it be possible to make TS have a "default route" that simply routes everything back out the actual default route?

I don't know how any of this works in detail but any VPN app must have some path to send packets so that they go "behind" the default route which routes most other apps into the VPN interface such that they just go out via the physical interface. Normally, VPN apps would encapsulate the packets in some sort of tunnel protocol and send the tunnel's packets out but wouldn't it then be possible to just ..not encapsulate? Just send out any packet that is not destined for TS addresses out the actual default route unmodified?

**Comment by ignoramous** [comment]
> Just send out any packet that is not destined for TS addresses out the actual default route unmodified?

I imagine on *nix, this requires CAP_NET_RAW (which isn't available on Android, for instance; but could use gvisor/netstack to "socksify" outgoing TCP & UDP, though).

**Comment by Kellwaxx11** [comment]
is it possible as a workaround to add an option to Tailscale android app, to automatically turn off Private DNS once tailscale is connected, and vise versa to turn it back on once tailscale disconnected.

might be possible using third-party apps like MacroDroid, Tasker, or Automate, but it would be nice to have it within Tailscale app

**Comment by JasonHK** [comment]
> is it possible as a workaround to add an option to Tailscale android app, to automatically turn off Private DNS once tailscale is connected, and vise versa to turn it back on once tailscale disconnected.
> 
> might be possible using third-party apps like MacroDroid, Tasker, or Automate, but it would be nice to have it within Tailscale app

@Kellwaxx11 Yes it's possible, but would (edit: very) likely be rejected by Google Play
